jumanji/environments/combinatorial/cvrp/env_test.py

```diff
@@ -0,0 +1,305 @@
+    state = new_state
+
+    new_state, next_timestep = step_fn(state, new_action)
+
+    # Check that the state has not changed
+    assert jnp.array_equal(new_state.position, state.position)
+    assert jnp.array_equal(new_state.visited_mask, state.visited_mask)
+    assert jnp.array_equal(new_state.order, state.order)
+    assert jnp.array_equal(new_state.num_total_visits, state.num_total_visits)
+
+
+def test_cvrp__does_not_smoke(cvrp_env: CVRP, capsys: pytest.CaptureFixture) -> None:
+    """Test that we can run an episode without any errors."""
+    check_env_does_not_smoke(cvrp_env)
+
+
+def test_tsp__trajectory_action(cvrp_env: CVRP) -> None:
+    """
+    Tests a trajectory by visiting nodes in increasing and cyclic order, visiting the depot when the next node in the
+    list surpasses the current capacity of the agent.
+    """
+    key = random.PRNGKey(0)
+    state, timestep = cvrp_env.reset(key)
+
+    # The position from which we will continue visiting after we recharge capacity at the depot.
+    pending_position = None
+    # actions = jnp.arange(1, cvrp_env.problem_size + 1)
+    while not timestep.last():
+        # Check that there are nodes that have not been selected yet.
+        assert state.visited_mask.sum() < (cvrp_env.problem_size + 1)
+
+        # Check that the reward is 0 while the trajectory is not done.
+        assert timestep.reward == 0
+
+        if pending_position is not None:
+            next_position = pending_position
+            pending_position = None
+        else:
+            # Select the next position. If we don't have enough capacity, set it
+            # as pending and go to the depot (0) to recharge.
+            next_position = (state.position % cvrp_env.problem_size) + 1
+            if state.capacity < state.problem[next_position, 2]:
+                pending_position = next_position
+                next_position = DEPOT_IDX
+
+        # If all cities have been visited, go to depot to finish the episode.
+        if state.visited_mask[1:].sum() == cvrp_env.problem_size:
+            next_position = DEPOT_IDX
+
+        state, timestep = cvrp_env.step(state, next_position)
+
+    # Check that the reward is negative when the trajectory is done.
+    assert timestep.reward < 0
+
+    # Check that no action can be taken (all nodes have been selected).
+    assert state.visited_mask.sum() == cvrp_env.problem_size + 1
+
+    assert timestep.last()
+
+
+def test_cvrp__invalid_revisit_node(cvrp_env: CVRP) -> None:
+    """Checks that an invalid action leads to a termination and the appropriate reward is received."""
+    key = random.PRNGKey(0)
+    state, timestep = cvrp_env.reset(key)
+
+    first_position = state.position
+    actions = (
+        jnp.array([first_position + 1, first_position + 2, first_position + 2])
+        % cvrp_env.problem_size
+    )
+
+    for a in actions:
+        assert timestep.reward == 0
+        assert timestep.step_type < StepType.LAST
```

Comment:
We provide `first`, `mid` and `last` methods for that.
```suggestion
        assert not timestep.last()
```